### PR TITLE
feat(rate-limit-upstream): integrate middleware with API

### LINKED ISSUE #28: [Feat] Add security measures to prevent DDOS attacks
## Issue Type

- [X] Bug report
- [X] Feature request
- [ ] Documentation update
- [ ] Other (please specify)

## Description

The repository doesn't have any security measures to prevent DDOS attacks and similar issues. 

## Additional context
[FastAPI Security 100 LVL: Production-Grade DDoS Protection](https://dev.to/adiletakmatov/fastapi-security-100-lvl-production-grade-ddos-protection-162k)

## Possible Solution

Adding a rate limiter to the HTTP client request on the user side. For example 2 calls per minute or 50 calls per hour. A rate limit that still allows for some testing and basic usage without being to over bearing, but limits mass spam calls to the FastAPI server.

### PR DESCRIPTION
**Pull request by MonoChromatical**
## Description

Adds a basic RateLimitMiddleware implementation for FastAPI requests. The middleware tracks request timestamps per client IP address and returns 429 Too Many Requests when the configured request limit is exceeded within the configured time window.

Also adds a focused tests covering the rate limit behavior.

Since this is also new territory, I've added heavy commenting and kept my original pseudocode for those who wish to read through it all for the first time and understand. (Later down the road when we expand this feature we can reduce my original commenting)

The middleware is now added to the main FastAPI app so the rate limiter runs as part of the API request flow.

The rate limit tests also cover API integration, the `Retry-After` response header, and the missing-client-IP safety path.

The README has been updated with a short summary of the rate limiter behavior.

AI was used just to help me fix Syncing issues I was encountering (The reason my branch has the prefix 'codex/' and was used to help assist me in working out how to implement the test for testing "none" IP address.

Included bellow is the transcripts in a drop down 

<details>
<summary>Screenshots / Transcript</summary>

<br>

<strong>Screenshot 1</strong>

<br>

<br><br>

<strong>Screenshot 2</strong>

<br>

<br><br>

<strong>Screenshot 3</strong>

<br>

</details>

## Type of change

Please delete options that are not relevant.

- [x] New feature 

## How Has This Been Tested?

Added `tests/mcp/test_rate_limit.py`, which verifies the rate limiter allows requests up to the configured limit, blocks excess requests with status code `429`, handles missing client IPs with status code `400`, and includes API integration coverage.

image

The test creates a small FastAPI app, attaches `RateLimitMiddleware`, and uses `TestClient` to send requests through the middleware. This tests the FastAPI request/response middleware flow without needing to manually start the live server.

The production rate limit is currently set to 50 requests per 3600 seconds. To avoid sending 51 requests during the test, I used `monkeypatch` to temporarily change `RATE_LIMIT_REQUESTS` to `2` and `RATE_LIMIT_WINDOW_SECONDS` to `60` for the duration of the test only. Pytest restores the original values after the test finishes.

I did not test this by running the full live server with `python -m main` because `monkeypatch` only affects the Python process running the test. A live server runs in a separate Python process, so changing `RATE_LIMIT_REQUESTS` inside pytest would not change the value used by the already-running server. Testing against the live server would require sending 51 real requests to hit the production limit.

I also added tests for the missing-client-IP safety path and API integration. The missing-client-IP test checks that if request.client is None, the middleware stops the request and returns 400 Bad Request instead of continuing with an "unknown" IP value.

This second test led to a bit of a rabbit hole because it could not be tested cleanly with `TestClient`. `TestClient` creates normal test requests that include client information, so it does not naturally hit the `request.client is None` branch. To test that specific edge case, the test calls `RateLimitMiddleware.dispatch()` directly with a minimal request-like object where `client=None`.

Because `dispatch()` is async and `BaseHTTPMiddleware` expects an ASGI app when the middleware is created, the test includes a small empty ASGI app and runs the async dispatch check with `asyncio.run()`. This keeps the test focused on the missing-client-IP behavior without needing to start the full server.

All tests have passed

## Checklist:

- [x] I have followed the coding style guidelines of this project
- [x] I have reviewed my own code
- [x] I have commented hard-to-understand areas of my code
- [x] I have added docstrings to all public functions/methods
- [x] I have used descriptive variable names to make my code easier to understand
- [x] I have made corresponding changes to the documentation
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] New and existing unit tests pass locally with my changes

## Related Issue(s)

Closes #28

### CODE REVIEW

**Comment by MonoChromatical** [comment]
Actually, since mentioning about the unexpected keyword argument id when running MCPTestClient
Do you want me to create an issue about the bug?

**Comment by Robbo-lab** [comment]
> Actually, since mentioning about the unexpected keyword argument id when running MCPTestClient Do you want me to create an issue about the bug?

Yes please

**Review by Robbo-lab — CHANGES_REQUESTED** [review]
Can you 

1. add test the integrate with the API 
2. add documentation in the README to summarise the rate limiter
3. can you update the title of the PR to match the format of the others

File: app/security/rate_limit.py [hunk 1]
```diff
@@ -0,0 +1,55 @@
...
+# BaseHTTPMiddleware is the class we inherit from to create custom middleware in FastAPI/Starlette.
+
+
+
+RATE_LIMIT_REQUESTS = 50
+RATE_LIMIT_WINDOW_SECONDS = 3600
+
+# Stores recent request timestamps for each client IP address.
+# defaultdict(deque) means a new IP automatically starts with an empty queue.
+requests_by_ip: dict[str, deque[float]] = defaultdict(deque)
```
Comment (Robbo-lab) [inline app/security/rate_limit.py:20]: Make sure to clear this out when before adding the new client ip this is a global variable

File: app/security/rate_limit.py [hunk 2]
```diff
@@ -0,0 +1,55 @@
...
+
+
+class RateLimitMiddleware(BaseHTTPMiddleware):
```
Comment (Robbo-lab) [inline app/security/rate_limit.py:23]: Can you add the middelware into the app app.add_middleware(RateLimitMiddleware) in main.py so your new limiter is working in the api (make sure to test it)

**Review by Robbo-lab — CHANGES_REQUESTED** [review]
See in code addiotnal comments, this should be good enough impliement in the app once done

File: app/security/rate_limit.py (hunk 1 above)
Comment (MonoChromatical) [reply app/security/rate_limit.py:20]: Image

Made a helper to implement this

File: app/security/rate_limit.py (hunk 2 above)
Comment (MonoChromatical) [reply app/security/rate_limit.py:23]: I can't believe I forgot this. I swear I did it previously. 
Thanks for catching that

**Review by Robbo-lab — APPROVED** [review]
